Fixed memory leaks caused by conditional free'ing for some TLS connec…

utoni opened the pull request:
…tions.

 * Members of tls_quic struct should also free'd if the detected master protocol is IMAPS / POPS / SMTPS / etc.

Signed-off-by: Toni Uhlig <matzeton@googlemail.com>

IvanNardi commented:
I was working  on a similar patch, since I saw the leak too.
The description is correct but I don't know if the fix is correct: tls_quic_stun struct is a member of an union. I think we need to be careful to free only valid pointers.

utoni commented:
Dangling pointer is indeed a critical issue and there are some possible double/invalid free's in nDPI e.g. `kerberos_buf.pktbuf`, 

There are multiple ways to prevent this. Use` memset()` in `ndpi_flow_malloc`, Set all dynamically allocated pointers in the flow struct to `NULL` after flow allocation (similar to `ndpi_reset_packet_line_info`).

We could also set the (not) correct subprotocol (TLS) which should also work and we wouldn't have to add protocol conditionals for SMTPS / POPS / IMAPS / etc. But for me it is irrational that TLS can be a subprotocol of SMTPS. It just makes more sense vice versa as SMTPS is the payload ("subprotocol") of TLS.

IvanNardi commented:
Another option might be to move "struct tls_quic_stun" out of the "union protos" (like we already did for most of http fields). Drawback: memory usage because "struct ndpi_flow_struct" will be bigger.
No idea which solution is the best one...

utoni commented:
I would love to see that `struct ndpi_flow_struct` doesn't get more bloated (if there there is another solution available).
For now a `sizeof(struct ndpi_flow_struct)` returns **2512 bytes**.
As that value might not be the effective memory used for all struct members due to memory alignment and resulting "holes". But it gives an insight about memory usage per flow. And especially for embedded devices, which usually suffer from random access memory shortage, that might be a killer for small embedded routers at some point.

utoni committed:
Fixed memory leaks caused by conditional free'ing for some TLS connections.

 * Members of tls_quic struct should also free'd if the detected master protocol is IMAPS / POPS / SMTPS / etc.

Signed-off-by: Toni Uhlig <matzeton@googlemail.com>

utoni committed:
Prevent reader_util.c from exit()'ing if maximum flow count reached.
This confuses the fuzzer.

 * Improved fuzz/Makefile.am to use LDADD for ../example/libndpiReader.a instead of LDFLAGS.
   That way, fuzz_ndpi_reader re-links to ../example/libndpiReader.a if something changed there.

Signed-off-by: Toni Uhlig <matzeton@googlemail.com>

utoni commented:
So I've decided to go with the protocol conditional spam and added another fix for `fuzz_ndpi_reader`.
